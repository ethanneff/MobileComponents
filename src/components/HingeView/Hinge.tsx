import React from "react";
import { Circle } from "react-native-svg";
import { useBoneMode } from "../../commons/Hooks/useBoneMode";
import { Point2 } from "../../commons/Utils/Maths/geometry/Point";

interface HingeProps {
  position: Point2;
  id?: string;
  strokeColor: string;
  strokeWidth?: number;
  fillColor: string;
  radius: number;
}

export default function Hinge({
  strokeWidth,
  strokeColor,
  fillColor,
  radius,
  position,
  id
}: HingeProps) {
  const boneMode = useBoneMode();
  if (boneMode) {
    return null;
  }
  return (
    <Circle
      id={id}
      cx={position.x}
      cy={position.y}
      r={radius}
      stroke={strokeColor}
      strokeWidth={strokeWidth}
      fill={fillColor}
    />
  );
}
