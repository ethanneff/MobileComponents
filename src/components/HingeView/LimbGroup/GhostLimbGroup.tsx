import React from "react";
import { Circle, ClipPath, Defs, G, Rect } from "react-native-svg";
import Victor from "victor";
import Theme from "../../../commons/Theme";
import { radiansToDegrees } from "../../../commons/Utils/Trigo";
import Hinge from "../Hinge";

interface GhostLimbGroupProps {
  startPosition: Victor;
  endPosition: Victor;
  scaledDown?: boolean;
  opacity?: number;
}

const styles = (scaledDown?: boolean) => ({
  ghostLimb: {
    strokeColor: Theme.color.primary200,
    strokeWidth: scaledDown ? 19 : 27
  },
  ghostHinge: {
    fillColor: Theme.color.primary200,
    strokeColor: Theme.color.primary200,
    radius: scaledDown ? 20 : 24
  }
});

export default function GhostLimbGroup({
  startPosition,
  endPosition,
  scaledDown,
  opacity
}: GhostLimbGroupProps) {
  const deltaX = endPosition.x - startPosition.x;
  const deltaY = endPosition.y - startPosition.y;
  const length = Math.sqrt(deltaX * deltaX + deltaY * deltaY);
  const radians = Math.atan2(deltaY, deltaX);
  const limbWidth = styles(scaledDown).ghostLimb.strokeWidth;
  const yOffset = limbWidth / 2;
  return (
    <>
      <Defs>
        <ClipPath id="clip">
          <G x={startPosition.x} y={startPosition.y}>
            <Rect
              y={-yOffset}
              width={length}
              height={limbWidth}
              fill={styles(scaledDown).ghostLimb.strokeColor}
              transform={`rotate(${radiansToDegrees(radians)})`}
            />
          </G>
          <Hinge
            position={endPosition}
            radius={styles(scaledDown).ghostHinge.radius}
            strokeColor={styles(scaledDown).ghostHinge.strokeColor}
            fillColor={styles(scaledDown).ghostHinge.fillColor}
          />
        </ClipPath>
      </Defs>
      <Circle
        cx={startPosition.x}
        cy={startPosition.y}
        opacity={opacity}
        r={length + styles(scaledDown).ghostHinge.radius}
        fill={styles(scaledDown).ghostLimb.strokeColor}
        clipPath="url(#clip)"
      />
    </>
  );
}
