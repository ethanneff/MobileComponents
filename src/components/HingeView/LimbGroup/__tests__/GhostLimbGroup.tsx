import React from "react";
import { Rect } from "react-native-svg";
import Victor from "victor";
import { testRender } from "../../../../commons/Utils/TestMocks/test-render";
import Hinge from "../../Hinge";
import GhostLimbGroup from "../GhostLimbGroup";

describe("GhostLimbGroup Component", () => {
  it("renders correctly", () => {
    const { tree } = testRender(
      <GhostLimbGroup
        startPosition={new Victor(10, 20)}
        endPosition={new Victor(20, 20)}
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("renders alternate styling when scaledDown is true", () => {
    const { findByType } = testRender(
      <GhostLimbGroup
        startPosition={new Victor(10, 20)}
        endPosition={new Victor(20, 20)}
        scaledDown
      />
    );

    expect(findByType(Rect).props.height).toBe(19);
    expect(findByType(Hinge).props.radius).toBe(20);
  });

  it("accepts a custom opacity", () => {
    const { findByProps } = testRender(
      <GhostLimbGroup
        startPosition={new Victor(10, 20)}
        endPosition={new Victor(20, 20)}
        opacity={0.25}
      />
    );

    expect(findByProps({ opacity: 0.25 })).toBeTruthy();
  });
});
