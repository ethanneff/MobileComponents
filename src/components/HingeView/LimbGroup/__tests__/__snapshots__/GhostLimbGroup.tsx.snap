// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GhostLimbGroup Component renders correctly 1`] = `
Array [
  <RNSVGDefs>
    <RNSVGClipPath
      fill={4278190080}
      fillOpacity={1}
      fillRule={1}
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
      name="clip"
      opacity={1}
      originX={0}
      originY={0}
      propList={Array []}
      rotation={0}
      scaleX={1}
      scaleY={1}
      skewX={0}
      skewY={0}
      stroke={null}
      strokeDasharray={null}
      strokeDashoffset={null}
      strokeLinecap={0}
      strokeLinejoin={0}
      strokeMiterlimit={4}
      strokeOpacity={1}
      strokeWidth={1}
      vectorEffect={0}
      x={0}
      y={0}
    >
      <RNSVGGroup
        fill={4278190080}
        fillOpacity={1}
        fillRule={1}
        font={Object {}}
        matrix={
          Array [
            1,
            0,
            -0,
            1,
            10,
            20,
          ]
        }
        opacity={1}
        originX={0}
        originY={0}
        propList={Array []}
        rotation={0}
        scaleX={1}
        scaleY={1}
        skewX={0}
        skewY={0}
        stroke={null}
        strokeDasharray={null}
        strokeDashoffset={null}
        strokeLinecap={0}
        strokeLinejoin={0}
        strokeMiterlimit={4}
        strokeOpacity={1}
        strokeWidth={1}
        vectorEffect={0}
        x={10}
        y={20}
      >
        <RNSVGRect
          fill={4286242010}
          fillOpacity={1}
          fillRule={1}
          height={27}
          matrix={
            Array [
              1,
              0,
              0,
              1,
              0,
              0,
            ]
          }
          opacity={1}
          originX={0}
          originY={0}
          propList={
            Array [
              "fill",
            ]
          }
          rotation={0}
          scaleX={1}
          scaleY={1}
          skewX={0}
          skewY={0}
          stroke={null}
          strokeDasharray={null}
          strokeDashoffset={null}
          strokeLinecap={0}
          strokeLinejoin={0}
          strokeMiterlimit={4}
          strokeOpacity={1}
          strokeWidth={1}
          vectorEffect={0}
          width={10}
          x={0}
          y={-13.5}
        />
      </RNSVGGroup>
      <RNSVGCircle
        cx={20}
        cy={20}
        fill={4286242010}
        fillOpacity={1}
        fillRule={1}
        matrix={
          Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ]
        }
        opacity={1}
        originX={0}
        originY={0}
        propList={
          Array [
            "fill",
            "stroke",
          ]
        }
        r={24}
        rotation={0}
        scaleX={1}
        scaleY={1}
        skewX={0}
        skewY={0}
        stroke={4286242010}
        strokeDasharray={null}
        strokeDashoffset={null}
        strokeLinecap={0}
        strokeLinejoin={0}
        strokeMiterlimit={4}
        strokeOpacity={1}
        strokeWidth={1}
        vectorEffect={0}
        x={0}
        y={0}
      />
    </RNSVGClipPath>
  </RNSVGDefs>,
  <RNSVGCircle
    clipPath="clip"
    cx={10}
    cy={20}
    fill={4286242010}
    fillOpacity={1}
    fillRule={1}
    matrix={
      Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ]
    }
    opacity={1}
    originX={0}
    originY={0}
    propList={
      Array [
        "fill",
      ]
    }
    r={34}
    rotation={0}
    scaleX={1}
    scaleY={1}
    skewX={0}
    skewY={0}
    stroke={null}
    strokeDasharray={null}
    strokeDashoffset={null}
    strokeLinecap={0}
    strokeLinejoin={0}
    strokeMiterlimit={4}
    strokeOpacity={1}
    strokeWidth={1}
    vectorEffect={0}
    x={0}
    y={0}
  />,
]
`;
